### DISCUSSION TITLE
Help writing correct grammar

### DISCUSSION THREAD

**Discussion by 02JanDal** [pest-parser/pest#953] [Q&A]
TL;DR: How do I get [this](https://pest.rs/?g=N4Ig5gTghgtjURALhAdQBIEkAqBRAygAoCCAwrgAQC8FA%2BsBQDohMsA%2BrjjALsxRwDlcqADKYhFAL6MAdrIDO3CAEsZYahQAkDZl10sAfhVUyAphApHdXEH2kyT5ihoACDAMYALBACopsrwRnDQpgWWCKAEIACmt9fk49ZgBKSwpiAQBNcIS4mzTY2xs%2BDjyS1gB6cuYAI2qQADN6uXZWCHreEGSc0qK9QwpCgFc%2BI2J8UkxMWnRcAA1aABFMAHEcYAAWSW6Ze1luKDBaUwAPKHduCKvr1wZo8cnp4hFCdAyAVQBZXAAlTFJciBaPUkCkANQJQr9UaDB5TWjPV4fb5-AG9UGtZjAroQqx9WxdfwyA7qG5k640UIUEnHM4XBIQUxgU5EgBuUAANkNTLQZEMYDUnOTgjQ3OkJvDlmtsBD7OyuTzlDAAA4c5TuZTcWiKFRqCKihhwp4vN4CL6-f6y2Ty7nC4WUhg2xUqtUarU61TqDgevX2RnMk528kOyow6KRZhVAYZTLJPx4qNEkTYIN2kPMAA8dlkycoqbJ6ZAGao2ZkKxT%2BYLzh0IAAfKXy3nK1dC7WSyx7O4APYyAAmmuUPdoSiG3E8AE9g9XqYc0swAPyl7t9gdDhqc%2BSTqdUklzkDziMdgI9-vcQcyWj%2Blnbhi7vFUAwwq%2BBzsn1cXsxgKDcUy9y9M68QzvVhIkfAZnyJZdT3PWgYFMAVzHkTxlGVFtp2A5gDHbNInUgt8zyHeQhhqH1SX1dDZzxHxsKMXDXxXAiL2VRkGmUQMqx3SjWAAPRoig6OPBiYOVLtFFY9iKQo9Q8U0PiBJkKD32OABHIZOU1Lc0M46TWDkzlbXo6Chw5H8gyArjMz0hVeX5QULEMpSTNTcydMzGEnRshD7MEoyLzAUy0ykvc23c-SeT5Ly8KEod-OcoK8XrAYPIiuyot82hVFMVT1O4TTnBcvdIisgyfKU5jTHkUwFNMG4CocxjaCgEiqvcGraunZhDzSXd6uExlKuq-8AxCBgIN6ocmoG1qhsAjqQC6owxtKhqfwgFRuC7CA8vy6dClUeRlF7Ux6i7UcDqOlI90KNJSL3bojwU-CYPcDku3cABrAB3ZRKskqlmB%2B2gXrer6ftMaJ7rSpSoGJZQgdej7vt%2BwtAZhs9gcRsGIaXJ7CNMbgzz1dr-pASqCc9a6jFuvFIfGj8u3dfHCbIkaQLJ5nKYoanWFp5aYMqjlTAuTbibCGQImYABtMNFIa9nPQSWWYJkBntSZhWOCVodVvWzbJ013GLwx0HfoN6KLzRuHjaRmqzfS9cOU3RXDeHCBRwnZ3zZmwM7aUz9v1-b3PfSuCvKQlDg6UoiSKUDWKC1piWLYyOGpEsTk99hqTNt%2BOXaczOYP8nOE9oWKC6HTLsrVXKU5grK1Or-Xc698qpuLl3Jpa9uvc7waIPLxOKq773UjxABdXRdlkEAABoQFUZVR2QEBJf62OLnPKgVdoQWGi1bghggGQx5ASQgA#editor) working

I'm trying to write a Pest grammar to parse MapCSS (no definitive formal specification, but most information can be found [here](https://josm.openstreetmap.de/wiki/Help/Styles/MapCSSImplementation)), and I'm struggling with a part of grammar.

Part of the syntax is that you can write conditions such as `[building=residential]` or `[width<10]` (similar to attribute selectors in normal CSS). I have a list of such selectors in the grammar (all looking something like `condition_equality = { tag ~ "=" ~ value }` but with different operators and sometimes a different right side sub-rule), but based on the error I get not all conditions are considered.

The relevant part of the code is [here](https://pest.rs/?g=N4Ig5gTghgtjURALhAdQBIEkAqBRAygAoCCAwrgAQC8FA%2BsBQDohMsA%2BrjjALsxRwDlcqADKYhFAL6MAdrIDO3CAEsZYahQAkDZl10sAfhVUyAphApHdXEH2kyT5ihoACDAMYALBACopsrwRnDQpgWWCKAEIACmt9fk49ZgBKSwpiAQBNcIS4mzTY2xs%2BDjyS1gB6cuYAI2qQADN6uXZWCHreEGSc0qK9QwpCgFc%2BI2J8UkxMWnRcAA1aABFMAHEcYAAWSW6Ze1luKDBaUwAPKHduCKvr1wZo8cnp4hFCdAyAVQBZXAAlTFJciBaPUkCkANQJQr9UaDB5TWjPV4fb5-AG9UGtZjAroQqx9WxdfwyA7qG5k640UIUEnHM4XBIQUxgU5EgBuUAANkNTLQZEMYDUnOTgjQ3OkJvDlmtsBD7OyuTzlDAAA4c5TuZTcWiKFRqCKihhwp4vN4CL6-f6y2Ty7nC4WUhg2xUqtUarU61TqDgevX2RnMk528kOyow6KRZhVAYZTLJPx4qNEkTYIN2kPMAA8dlkycoqbJ6ZAGao2ZkKxT%2BYLzh0IAAfKXy3nK1dC7WSyx7O4APYyAAmmuUPdoSiG3E8AE9g9XqYc0swAPyl7t9gdDhqc%2BSTqdUklzkDziMdgI9-vcQcyWj%2Blnbhi7vFUAwwq%2BBzsn1cXsxgKDcUy9y9M68QzvVhIkfAZnyJZdT3PWgYFMAVzHkTxlGVFtp2A5gDHbNInUgt8zyHeQhhqH1SX1dDZzxHxsKMXDXxXAiL2VRkGmUQMqx3SjWAAPRoig6OPBiYOVLtFFY9iKQo9Q8U0PiBJkKD32OABHIZOU1Lc0M46TWDkzlbXo6Chw5H8gyArjMz0hVeX5QULEMpSTNTcydMzGEnRshD7MEoyLzAUy0ykvc23c-SeT5Ly8KEod-OcoK8XrAYPIiuyot82hVFMVT1O4TTnBcvdIisgyfKU5jTHkUwFNMG4CocxjaCgEiqvcGraunZhDzSXd6uExlKuq-8AxCBgIN6ocmoG1qhsAjqQC6owxtKhqfwgFRuC7CA8vy6dClUeRlF7Ux6i7UcDqOlI90KNJSL3bojwU-CYPcDku3cABrAB3ZRKskqlmB%2B2gXrer6ftMaJ7rSpSoGJZQgdej7vt%2BwtAZhs9gcRsGIaXJ7CNMbgzz1dr-pASqCc9a6jFuvFIfGj8u3dfHCbIkaQLJ5nKYoanWFp5aYMqjlTAuTbibCGQImYABtMNFIa9nPQSWWYJkBntSZhWOCVodVvWzbJ013GLwx0HfoN6KLzRuHjaRmqzfS9cOU3RXDeHCBRwnZ3zZmwM7aUz9v1-b3PfSuCvKQlDg6UoiSKUDWKC1piWLYyOGpEsTk99hqTNt%2BOXaczOYP8nOE9oWKC6HTLsrVXKU5grK1Or-Xc698qpuLl3Jpa9uvc7waIPLxOKq773UjxABdXRdlkEAABoQFUZVR2QEBJf62OLnPKgVdoQWGi1bghggGQx5ASQgA#editor)

**Comment by tomtau** [comment]
Which one is the top-level rule to run? One thing I can't see in the grammar is something that would match `[` and `]` -- when I remove them (i.e. the input string is `restriction=no_left_turn`) and select the `condition_equality` rule, it seems to work as intended?

**Reply by 02JanDal** [reply]
The `condition_selector` rule is the "top level rule", it's a pair of brackets and then a choice between all the other `condition_*` rules. The fact that `condition_equality` indeed does match the contents in the brackets just makes it even weirder to me, especially as `condition_equality` is not even mentioned as an option at that location (while for example `condition_inequality` is mentioned, despite the only difference being the operator):

```
 --> 1:2
  |
1 | [restriction=no_left_turn]
  |  ^---
  |
  = expected condition_anti_clockwise, condition_clockwise, condition_falsy, condition_gt, condition_gte, condition_inequality, condition_lt, condition_lte, condition_membership, condition_negated_regex, condition_not_setting, condition_postfix, condition_prefix, condition_regex, condition_setting, condition_substring, condition_territory, or condition_truthy
```
